Repository and image quota support

miminar opened the pull request:
New resource quotas:
    
1. openshift.io/projectimagessize
2. openshift.io/imagestreamsize
3. openshift.io/imagesize

The 1st allows to set a limit for a sum of sizes of all image layers
accross all repositories in a namespace.

The 2nd allows to set a limit for a sum of sizes of all image layers
in a single repository. Usage will always be 0 because we cannot store
different values for different repositories.

The 3rd allows to set a limit for a sum of sizes of all layers of
single image. Usage will always be 0.

Layers occuring multiple times in a single repository count just once.

**TODO**:

- [x] write godocs
- [x] avoid patching upstream `GenericEvaluator`, write our own instead
- [x] fix tests
- [x] write tests

miminar commented:
cc @kargakis, @soltysh, @derekwaynecarr

derekwaynecarr commented:
I actually did not find the names bad, but we can't patch upstream to allow non-namespace resource prefixes.

derekwaynecarr commented:
fwiw, I rebased the PR this builds on: https://github.com/kubernetes/kubernetes/pull/20446

derekwaynecarr commented:
A number of questions, but for the most part, I think this validates I didn't waste my time with the upstream refactor.

soltysh commented:
@miminar I did first pass, it generally looks good, a few comments here and there, but most importantly the question that pops in my head after going through what about the origin update use-case, where you already have existing images. From looking at the code I see you'll apply the quota only to newly created images, is that intentional? Do we expect some migration/update tool to handle that?

miminar commented:
@soltysh I doubt anyone would like resourcequota controller start deleting existing images when he updates or even when he set the quotas on existing project. Even if wanted to handle the case, how would we choose the images to delete?

I'd rather just let admins list images/repositories exceeding quotas and let them decide. We could also extend `oadm prune images` to recognize images exceeding `openshift.io/imagesize` quota.

That's something to do in a follow-up though.

miminar commented:
- rebased on newer 20446
- renamed `totalimagesize` to `projectimagessize`
- renamed `repositorysize` to `imagestreamsize`
- added extended test
- fixes

miminar commented:
> A number of questions, but for the most part, I think this validates I didn't waste my time with the upstream refactor.

@derekwaynecarr you certainly did not, thanks for the great job and suggestions.

miminar commented on pkg/dockerregistry/server/quotarestrictedblobstore.go:
```diff
@@ -0,0 +1,155 @@
+ *      1. Origin master calculates image sizes from the contents of the layers. Registry, on the other hand,
+ *         deals with layers themselves which contain some overhead required to store file attributes, and the
+ *         layers are compressed. Thus we compare apples with peers. The check is primarily useful when the
+ *         quota is already almost reached.
```
@ncdc can you please take a look? Does this makes sense?

ncdc commented on pkg/dockerregistry/server/quotarestrictedblobstore.go:
```diff
@@ -0,0 +1,155 @@
+ *
+ *      1. Origin master calculates image sizes from the contents of the layers. Registry, on the other hand,
+ *         deals with layers themselves which contain some overhead required to store file attributes, and the
+ *         layers are compressed. Thus we compare apples with peers. The check is primarily useful when the
```
peers -> pears?

miminar replied:
Of course! :)

ncdc commented on pkg/dockerregistry/server/quotarestrictedblobstore.go:
```diff
@@ -0,0 +1,155 @@
+ *         etcd, not dangling layers inside a registry. Thus, during a check, we take into account just the
+ *         single layer, not the image as a whole.
+ *
+ *         This leads to a situation where several layers can be written until a big enough will be received
```
until a big enough *layer*?

ncdc commented on pkg/dockerregistry/server/quotarestrictedblobstore.go:
```diff
@@ -0,0 +1,155 @@
+ *         that exceeds quota limit.
+ *
+ *      3. Image stream size quota doesn't accumulate. Iow, its usage is NOT stored in a resource quota object
+ *         (its always 0 (well almost - except for the very short period time noted in 2)). Therefore its
```
its always -> it's always

miminar replied:
Addressed, thanks!
On 02/12/2016 04:14 PM, Andy Goldstein wrote:
> In pkg/dockerregistry/server/quotarestrictedblobstore.go
> <https://github.com/openshift/origin/pull/7128#discussion_r52752757>:
>
>> + *      1. Origin master calculates image sizes from the contents of the layers. Registry, on the other hand,
>> + *         deals with layers themselves which contain some overhead required to store file attributes, and the
>> + *         layers are compressed. Thus we compare apples with peers. The check is primarily useful when the
>> + *         quota is already almost reached.
>> + *
>> + *      2. During a push, multiple layers are uploaded. Each uploaded layer will, for a very short period of
>> + *         time, raise quota usage. However, the usage will be shortly overwritten by a resource quota refresh
>> + *         back to the original value because the RQ controller takes into account just images stored in
>> + *         etcd, not dangling layers inside a registry. Thus, during a check, we take into account just the
>> + *         single layer, not the image as a whole.
>> + *
>> + *         This leads to a situation where several layers can be written until a big enough will be received
>> + *         that exceeds quota limit.
>> + *
>> + *      3. Image stream size quota doesn't accumulate. Iow, its usage is NOT stored in a resource quota object
>> + *         (its always 0 (well almost - except for the very short period time noted in 2)). Therefore its
>
> its always -> it's always
>
> —
> Reply to this email directly or view it on GitHub <https://github.com/openshift/origin/pull/7128/files#r52752757>.
>

ncdc commented on pkg/dockerregistry/server/quotarestrictedblobstore.go:
```diff
@@ -0,0 +1,155 @@
+
+// admitBlobWrite checks whether the blob of given size does not exceed image quota, if set. Returns
+// ErrorCodeDenied error if the quota is exceeded.
+func admitBlobWrite(ctx context.Context, repo *repository, size int64) error {
```
cc @derekwaynecarr for quota review

miminar commented:
[test]

miminar commented:
@soltysh could you give it another look, please?

soltysh commented:
On it...

soltysh commented:
This LGTM, fix tests.

Squash following commits into one:
* Update generated swagger-spec (https://github.com/miminar/origin/commit/c0f97d62d9d927fbb34a64db86929cee5a240d70)
* Added extended test for image quota (https://github.com/openshift/origin/commit/fd96e0db6d945de3cd792de724db4524b84e39ee)
* Resource quota for repositories and images (https://github.com/miminar/origin/commit/05bfee4f7f46ad3edfbc8d43b86b838a2e42fc95)

And these two into one as well (separate from previous):
* Allow registry to fetch resourcequota (https://github.com/miminar/origin/commit/b9433ea95fbdcd5914df5487066c9e05b3578254)
* Added additional quota check for layer upload in a registry (https://github.com/openshift/origin/commit/a831adf6e6473c08af28100981a2719582350c94)

miminar commented:
@soltysh Thanks! squashed. re-[test][extended:core(image resource quota)]

soltysh commented:
LGTM, although it looks like one of your extended tests is failing. 

This PR will be merged once #7269 lands.

miminar commented:
Hmm, it looks like Docker 1.8.2 rounds built layers to multiples of 2^10 B. I'll have to tune layer sizes.

ncdc commented:
Need to make sure things are good with 1.8.2, 1.9.x, 1.10.x if possible.

miminar commented:
[test][extended:core(image resource quota)]

miminar commented:
So the extended test is passing now, but there was also this flake:

    Error from server: error when replacing "STDIN": error #0: client: etcd member https://172.18.9.140:4001 has no leader
    !!! Error in /data/src/github.com/openshift/origin/hack/util.sh:592
	    'oc replace scc privileged -f - --config="${ADMIN_KUBECONFIG}"' exited with status 0 0 1
    Call stack:
	    1: /data/src/github.com/openshift/origin/hack/util.sh:592 install_router(...)
	    2: test/extended/core.sh:71 main(...)
    Exiting with status 1

In [job 10939](https://ci.openshift.redhat.com/jenkins/job/test_pull_requests_origin/10939/consoleFull).

miminar commented:
https://github.com/openshift/origin/pull/7269 is merged now. re-[test][extended:core(image resource quota)]

miminar commented:
[test][extended:core(image resource quota)]

miminar commented:
I wanted to confirm that this works also with [docker 1.10.1](https://github.com/projectatomic/docker/tree/rhel7-1.10.1) but unfortunately it's not the case. Newer docker no longer sets `Size` attribute in history of layers. So now there's no useful information in manifest we can use to calculcate image size.

To deal with that, we'd either have to quicly switch to [manifest schema 2-2](https://github.com/docker/distribution/blob/master/docs/spec/manifest-v2-2.md#manifest-list-field-descriptions) which embeds mandatory layer sizes. Or make the master server query registry for layer sizes if they are not available in the manifest. I'm more inclined to do the latter.

soltysh commented:
Discussed this with @miminar over IRC, we'll use the latter proposed solution for now, iow. where master query registry for layer sizes if not available. And a new card will be created to migrate to use schema 2-2.

miminar commented:
--- FAIL: TestClusterResourceOverridePlugin-2 (7.54s)
    	clusterresourceoverride_admission_test.go:69: limit-with-default: Memory req did not match expected 256Mi: 128Mi
    	clusterresourceoverride_admission_test.go:72: limit-with-default: CPU limit did not match expected 500 mcore: 1
    	clusterresourceoverride_admission_test.go:75: limit-with-default: CPU req did not match expected 250 mcore: 500m
    FAIL

Flake #7401

miminar commented:
[test][extended:core(image resource quota)]

miminar commented:
I consider it a successful run.

soltysh commented:
Yup, the other failed because of AWS prolem: `The instance never became "ready" in AWS. The timeout currently`

soltysh commented:
LGTM, waiting for @pweil- sing-off

miminar commented:
I'd merge this as is. I'm working on a follow-up to treat newer Docker versions. Docker 1.10 won't be in use for a while.

liggitt commented:
> waiting for @pweil- sing-off

:microphone:

pweil- commented:
> waiting for @pweil- sing-off

:smile: that is the last thing you want, trust me

soltysh commented:
:guitar: sounds better then

ncdc commented:
Any chance we can get those carry commits upstream?

miminar commented:
@ncdc I think there's a high chance. I'll try and submit them later.

ncdc commented:
We've learned our lesson the hard way more than once... Please don't merge this PR until the upstream PRs are merged.

liggitt commented:
> Please don't merge this PR until the upstream PRs are merged.

:+1:

miminar commented:
Upstream PRs are [#1473](https://github.com/docker/distribution/pull/1473) and [#1474](https://github.com/docker/distribution/pull/1474). Updated commit messages to reflect that.

miminar commented:
Rebased. I decided not to clutter this with a support for `docker 1.10`. I made a follow-up #7590 to address that. There are couple of important things to resolve with no easy solution known.

miminar commented:
Hmm. Docker 1.8.2 uploads much larger layers than 1.9. Need to fine tune layer sizes in the extended test again.

miminar commented:
Flake #6435 and one more which seems to be a new one.

miminar commented:
re-[test]

miminar commented:
Addressed comments except for the package godoc.

miminar commented:
Got LGTMs for docker/distribution PRs.

miminar commented:
Got successful run. And #7243 once again.

soltysh commented:
LGTM still
@liggitt @ncdc are you guys comfortable merging this knowing we have lgtm on all upstream prs?

ncdc commented:
I would still feel most comfortable if it actually merged

soltysh commented:
> I would still feel most comfortable if it actually merged

I'm not afraid about k8s PR, that should be merged soon, besides we'll be waiting with rebase for it. Not sure about docker ones :/. I'm ok with waiting but if all other dependencies for rebase will be met I'd say merge this one and proceed with rebase. /cc @liggitt

miminar commented:
2 out of 3 PRs merged. [Distribution PR#1473](https://github.com/docker/distribution/pull/1473) has just a LGTM.

pweil- commented:
Both upstreams have been merged now so unless there is any objection this should be ready for merge.

@miminar - can you double check that the commits in this branch are the same as what merged
@liggitt  - do you want to get this in before starting the rebase?

pweil- commented:
[test]

liggitt commented:
yeah, before rebase is fine. still holding on some large k8s prs

miminar commented:
> @miminar - can you double check that the commits in this branch are the same as what merged

@pweil- double-checked. No interesting deviation found in the rebased PR.

pweil- commented:
re[test]

pweil- commented:
Failing on the same issue as: https://github.com/openshift/origin/issues/7355

miminar commented:
That's related to #6335. I'm working on a retry functionality for the integration test suite. The pulp registry seems to be working fine again. re-[test].

soltysh commented:
Flake from https://github.com/openshift/origin/issues/7616
re-[test]

miminar commented:
Looks like a new flake #7729 in both runs. re-[test].

soltysh commented:
Flakes from #7662 and #7355, I wonder it it's possible to have clean run on this one :/ I will retest and merg after next tests if it get at least one clean run out of the two.

re-[test]

miminar commented:
Actually, there was a [SUCCESS run](https://ci.openshift.redhat.com/jenkins/job/test_pr_origin/1763/) at 10:11 AM CET :smile:

soltysh commented:
ALL GREEN!!!
[merge]

pweil- commented:
:tada: :cake:

soltysh commented:
Flake https://github.com/openshift/origin/issues/7734:
re-[merge]

soltysh commented:
:tada:

smarterclayton commented:
> The 1st allows to set a limit for a sum of sizes of all image layers all repositories in a namespace.

How is this enforced when images are tagged via image stream (imagestreamtag)?  We don't call admission controllers then.

miminar commented:
@smarterclayton do we want to count imagestreamtags at all? The layers are not created in the registry because the image isn't pushed. If we delete the image, it won't be available under any istag. We can't even pull an image manually from an istag:

    $ oc tag -n tmp tmp/hello-world@sha256:bb5c16ecd58633d4088140acb6ee386c5c3b09a8f03c517a288093415f0c09d1 mybase:latest
    $ docker pull 172.30.150.240:5000/tmp/mybase:latest
    Trying to pull repository 172.30.150.240:5000/tmp/mybase-world ... latest: Pulling from tmp/mybase
    b901d36b6f2f: Pulling fs layer 
    0a6ba66e537a: Pulling fs layer 
    not found
    Error: image tmp/mybase:latest not found

Currently the istag is being added to the project's usage although it doesn't go through the admission check. Which is clearly wrong. The question is do we want to count it and admit it or ignore it completely? I'm in favor of the latter.

Actually, wouldn't it be better to re-push the tagged image into the internal registry under a new name if we refer to an internally managed image?

smarterclayton commented:
The error message you are getting is unexpected.  What version of docker?

miminar commented:
`docker-1.9.1-16.el7.x86_64`

ncdc commented:
@miminar was that a copy/paste error? You asked docker to pull tmp/mybase:latest and it says it's pulling tmp/mybase-world?

miminar commented:
@ncdc no, I changed the repo names and missed one `-world` occurence. I'll re-paste without modification with docker 1.8.

ncdc commented:
It might also be worthwhile running docker in debug mode (-D) to capture more verbose output from the daemon.

miminar commented:
Instead of posting it here, see the [gist](https://gist.github.com/miminar/95cfda439355769f8f8f).

I don't understand, how the pull could succeed for a tag created with `oc tag` if the layers aren't linked in to the repository of registry's storage.

smarterclayton commented:
So today, I can work around quota by doing the following:

Upload image to project1
Tag image into project2
Delete image stream in project1
<repeat>

miminar commented:
@smarterclayton yes, but you won't be able to pull it.

smarterclayton commented:
You should be able to pull it - if not, that's a bug.

On Mar 3, 2016, at 10:53 AM, Michal Minar <notifications@github.com> wrote:

@smarterclayton <https://github.com/smarterclayton> yes, but you won't be
able to pull it.

—
Reply to this email directly or view it on GitHub
<https://github.com/openshift/origin/pull/7128#issuecomment-191824172>.
